numpy/lib/npyio.py

```diff
@@ -2058,24 +2057,27 @@ def encode_unicode_cols(row_tup):
 
             try:
                 data = [encode_unicode_cols(r) for r in data]
-                type_str = np.bytes_
             except UnicodeEncodeError:
                 pass
+            else:
+                for i in strcolidx:
+                    column_types[i] = np.bytes_
 
 
-        # ... and take the largest number of chars.
-        for i in strcolidx:
-            max_line_length = max(len(row[i]) for row in data)
-            column_types[i] = np.dtype((type_str, max_line_length))
+        # Update string types to be the right length
+        for i, col_type in enumerate(column_types):
+            if np.issubdtype(col_type, np.character):
+                column_types[i] = (col_type, max(len(row[i]) for row in data))
+
         #
         if names is None:
             # If the dtype is uniform, don't define names, else use ''
-            base = set([c.type for c in converters if c._checked])
+            base = set([
+                c_type
+                for c, c_type in zip(converters, column_types)
+                if c._checked])
```

Comment:
I have no idea what this is testing for, but these lines are just cleanup anyway, now that `typestr` is consistent with `column_types`